### PR TITLE
Implement k01 fr 06 validity period check

### PR DESCRIPTION
**Pull request by ghost**
## Checklist before requesting a review
- [ x ] I have performed a self-review of my code
- [ x ] I have made corresponding changes to the documentation
- [ x ] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [ x ] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: lib/ocpp/v201/smart_charging.cpp [hunk 1]
```diff
@@ -125,10 +131,14 @@ ProfileValidationResultEnum SmartChargingHandler::validate_tx_profile(const Char
...
  */
+
 ProfileValidationResultEnum
 SmartChargingHandler::validate_profile_schedules(ChargingProfile& profile,
                                                  std::optional<EvseInterface*> evse_opt) const {
-    for (ChargingSchedule& schedule : profile.chargingSchedule) {
+    auto schedules = profile.chargingSchedule;
+
+    for (auto& schedule : schedules) {
+
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: Why change this back? The Everest folks asked to do it in one step.

File: lib/ocpp/v201/smart_charging.cpp [hunk 2]
```diff
@@ -224,4 +232,86 @@ std::vector<ChargingProfile> SmartChargingHandler::get_station_wide_tx_default_p
...
+                        existing_profile.chargingProfileKind == candidate_profile.chargingProfileKind &&
+                        existing_profile.id != candidate_profile.id) {
+
+                        // Existing profile last forever should be updated to replace
+                        if (!existing_profile.validFrom.has_value() && !existing_profile.validTo.has_value()) {
+                            return true; // reject
+                        }
+
+                        // Cannot accept new profile that last forever when there is an existing profile
+                        if (!candidate_profile.validFrom.has_value() && !candidate_profile.validTo.has_value()) {
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: If the existing profile has a `validTo` that ends before the candidate is applied (which is the moment it's received in this case), shouldn't the candidate be acceptable?

**Review by christopher-davis-afs — COMMENTED** [review]
Overall this feels like a big block of logic with a lot of if statements, and I'm wondering: is there a way to simplify this into math?

File: lib/ocpp/v201/smart_charging.cpp (hunk 2 above)
Comment (ghost) [reply lib/ocpp/v201/smart_charging.cpp]: Thanks for the pairing session. Concerns are addressed.

**Review by christopher-davis-afs — APPROVED** [review]

**Review by gberardi-pillar — APPROVED** [review]

**Pull request merged**